### DISCUSSION TITLE
does rook-ceph  rgw limit the connection numbers

### DISCUSSION THREAD

**Discussion by wxcoin** [rook/rook#12888] [Q&A]
the java program through error below,does rook-ceph rgw limit the connection numbers,if there exist how to edit in rook-ceph

i have search a lot but didn't figure out.

#########################################################################################
at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
[tomcat-embed-core-9.0.33.iar!/:9.0.33]
at java.lang.Thread.run(Thread.java:750)~[?:1.8.0342]
Caused by: org.apache,http.conn.ConnectionPoolTimeoutException: Tineout waiting for connection from pool
PoolingHttpclientConectionllanager,leaseCornectin(PoolingHttpclientCornectionlanager.java:316)"[httpclient-4.5.12. jar!/:4.5.12)
at org.apache.http.impl.conn.
at org. apache.htp.impl.comn,PolingHttpclientomectionllanager$1.get(FoolingtpclientCmnectionllanager.java22)"httpclient-4.5.12.jar!/:4.5.12)
t sunreflect.Generatedhethoddccessor317.invokellnknowm Sourcel?:?
at sun.reflect.DelegatingllethodAccessorInpl.invoke(DelegatingmethodAccessorImol,java:43) "[?:1,8.0 342
at iavalang.reflect Methodinyoke(Method iava:498)“[?:1.803421
clientConnectionRequestFactory$Handler,invoke (ClientConmectionRequestFactory. java:70) “aws-ava-sdk-core-1,11,415.jar!/:?)
 
tProx252.getilnknow
Source)~[?:?]
at com.amazonaws.http.conr
execute(MainClientExec.java:190)~[httpclient-4.5.12.ar!/:4.512]
at org.apache
++
~[httpclient-4.5.12.jar!/:4.5.12]
xecutelProtocolExec.awa:18l
t org.
doExecute(InternalHttpClient.java:185)“[httpclient-4.5.12jar!/:4.5.12
 
.java:83)~[httpclient-4.5.12.jar!/:4.5.12]
cute(closeableHttpclient.java:56) ~[httpclient-4.5.12.jar!/:4.5.12]
t org.apache
clent.Closeableht
t com.amazonaws.http.apache.client.impl.SdkHttpClient.
execute(SdkHttpClient.java:72) ~[aws-java-sdk-core-1.11.415.jar!/:?]
t com.amazonaws.htto.AmazonH+toClienttReque
xecuteOneRequest(AmazonHttpClient.java:1258) “[aws-java-sdk-core-1.11.415.jar!/:?]
iaza:10741
~[aws-iava-sdk-core-1.11.415.jar!/:?]
 
102 more
2023-09-1114:57:47.068[http-nio-17771-exec-6] INFO
.hcki.assist.
document.provider.DocumentUploadProvider - document upload ceph error
com.amazonaws.SdkClientException: Unable to execute HTTP request:
waiting for connection from pool
 
![javapool](https://github.com/rook/rook/assets/83400374/aec13d45-b082-4449-bcc5-c1d0e78e6378)

at com amazonaws.htp.MnazonHtpclient$RequestExecutor,handlefetryableException anazonHttpclient.jaya:l136)"aws-jaa-sdk-core-1.1.415.jar!/:?)
at com, amnazonaws.htp.AmazonHttpclient$RequestExecutorexecutehelper AmazonHttpclient.java:1082) "aws-java-sdk-core-1.11.415.jar!/:?
at com.amazonaws.http.AmazonHttpClient$RequestExecutor
[e(AmazonHttpClient.java:745)~[aws-ava-sdk-core-1.11.415.jar!/:?][](url)](url)

**Comment by zhucan** [comment]
![image](https://github.com/rook/rook/assets/13997135/3e064836-f8ad-4eaf-b845-ac054b16406c)
like this?

**Reply by wxcoin** [reply]
rgw_thread_pool_size
Number of threads for the Beast server. Increase to a higher value if you need to serve more requests. Defaults to 100 threads.
############

it seems   Defaults to 100

edit in ceph.conf, how to make it permanently work in rook-ceph ?

vi /etc/ceph/ceph.conf
[global]
rgw_thread_pool_size = 5000

**Comment by zhucan** [comment]
![image](https://github.com/rook/rook/assets/13997135/7f6e3b88-e01c-447a-a97b-e662f70cf1da)
update the rook-config-override configmap and restart the rgw pod.

**Comment by wxcoin** [comment]
[root@k8s-master2 ~]# k get configmap -n rook-ceph 
NAME                                 DATA   AGE
csi-ceph-conf-override               1      22h
kube-root-ca.crt                     1      146d
local-device-k8s-master1             1      146d
local-device-k8s-master2             1      146d
local-device-k8s-master3             1      146d
local-device-k8s-node1               1      146d
rook-ceph-csi-config                 1      146d
rook-ceph-csi-mapping-config         1      146d
rook-ceph-mon-endpoints              5      146d
rook-ceph-operator-config            26     146d
rook-ceph-pdbstatemap                2      146d
rook-ceph-rgw-my-store-mime-types    1      146d
rook-ceph-rgw-pro-store-mime-types   1      97d
rook-config-override                 1      146d






[root@k8s-master2 ~]#  kubectl -n rook-ceph get pods -o wide
NAME                                                    READY   STATUS      RESTARTS        AGE     IP              NODE          NOMINATED NODE   READINESS GATES
csi-cephfsplugin-55wpc                                  2/2     Running     4 (127d ago)    146d    10.10.5.162     k8s-master3   <none>           <none>
csi-cephfsplugin-6x7wm                                  2/2     Running     16 (13d ago)    146d    10.10.5.160     k8s-master1   <none>           <none>
csi-cephfsplugin-gkfrh                                  2/2     Running     5 (107d ago)    132d    10.10.5.161     k8s-master2   <none>           <none>
csi-cephfsplugin-provisioner-75875b5887-7zgfj           5/5     Running     35 (107d ago)   146d    10.18.135.202   k8s-master3   <none>           <none>
csi-cephfsplugin-provisioner-75875b5887-rch9p           5/5     Running     0               13d     10.18.36.113    k8s-node1     <none>           <none>
csi-cephfsplugin-t9fbd                                  2/2     Running     4 (127d ago)    146d    10.10.5.163     k8s-node1     <none>           <none>
csi-rbdplugin-hc467                                     2/2     Running     4 (127d ago)    146d    10.10.5.162     k8s-master3   <none>           <none>
csi-rbdplugin-l226z                                     2/2     Running     4 (107d ago)    132d    10.10.5.161     k8s-master2   <none>           <none>
csi-rbdplugin-provisioner-56d69f5d8-9mjj4               5/5     Running     0               13d     10.18.224.31    k8s-master2   <none>           <none>
csi-rbdplugin-provisioner-56d69f5d8-9pvfz               5/5     Running     15 (107d ago)   139d    10.18.135.199   k8s-master3   <none>           <none>
csi-rbdplugin-sdpd5                                     2/2     Running     12 (13d ago)    146d    10.10.5.160     k8s-master1   <none>           <none>
csi-rbdplugin-t7q5d                                     2/2     Running     4 (127d ago)    146d    10.10.5.163     k8s-node1     <none>           <none>
rook-ceph-crashcollector-k8s-master1-77c58f7646-nb2jq   1/1     Running     0               29h     10.18.159.189   k8s-master1   <none>           <none>
rook-ceph-crashcollector-k8s-master2-5f5b5678d5-c69pg   1/1     Running     2 (107d ago)    127d    10.18.224.50    k8s-master2   <none>           <none>
rook-ceph-crashcollector-k8s-master3-d6f897ccd-6sgdh    1/1     Running     1 (127d ago)    139d    10.18.135.254   k8s-master3   <none>           <none>
rook-ceph-crashcollector-k8s-node1-67b5f7b7f6-j8rp9     1/1     Running     1 (127d ago)    136d    10.18.36.74     k8s-node1     <none>           <none>
rook-ceph-mds-myfs-a-6898fcb7fc-wzl72                   2/2     Running     5 (107d ago)    127d    10.18.224.12    k8s-master2   <none>           <none>
rook-ceph-mds-myfs-b-dffd4b8d6-9fjwp                    2/2     Running     3 (127d ago)    136d    10.18.36.106    k8s-node1     <none>           <none>
rook-ceph-mgr-a-699db58f86-wcrz6                        3/3     Running     0               13d     10.18.135.250   k8s-master3   <none>           <none>
rook-ceph-mgr-b-799f465469-cj4q6                        3/3     Running     5 (127d ago)    136d    10.18.36.105    k8s-node1     <none>           <none>
rook-ceph-mon-b-68556bb4dc-68pcm                        2/2     Running     4 (127d ago)    146d    10.18.135.200   k8s-master3   <none>           <none>
rook-ceph-mon-e-968779c54-7g67t                         2/2     Running     2 (127d ago)    136d    10.18.36.120    k8s-node1     <none>           <none>
rook-ceph-mon-f-ff8b578b4-nxv6j                         2/2     Running     0               13d     10.18.224.35    k8s-master2   <none>           <none>
rook-ceph-operator-7696fc6bd5-mgx7m                     1/1     Running     0               21h     10.18.224.48    k8s-master2   <none>           <none>
rook-ceph-osd-0-6d79bccd6f-qmr7q                        2/2     Running     0               13d     10.18.159.185   k8s-master1   <none>           <none>
rook-ceph-osd-2-7b5466bb5c-dm2zp                        2/2     Running     2 (127d ago)    139d    10.18.36.116    k8s-node1     <none>           <none>
rook-ceph-osd-3-776f8c5698-z5n4l                        2/2     Running     4 (127d ago)    146d    10.18.135.195   k8s-master3   <none>           <none>
rook-ceph-osd-4-769f5c8564-9gmsp                        2/2     Running     4 (107d ago)    132d    10.18.224.22    k8s-master2   <none>           <none>
rook-ceph-osd-prepare-k8s-master1-sc246                 0/1     Completed   0               8m35s   10.18.159.139   k8s-master1   <none>           <none>
rook-ceph-osd-prepare-k8s-master2-5rdnk                 0/1     Completed   0               8m32s   10.18.224.8     k8s-master2   <none>           <none>
rook-ceph-osd-prepare-k8s-master3-tgnx7                 0/1     Completed   0               8m29s   10.18.135.236   k8s-master3   <none>           <none>
rook-ceph-osd-prepare-k8s-node1-rd8hr                   0/1     Completed   0               8m26s   10.18.36.101    k8s-node1     <none>           <none>
rook-ceph-rgw-my-store-a-6b8bfd75c6-99p5n               2/2     Running     0               8m40s   10.18.36.93     k8s-node1     <none>           <none>
rook-ceph-rgw-pro-store-a-7cfbcb44f8-sw4sp              2/2     Running     0               23h     10.18.159.129   k8s-master1   <none>           <none>
rook-ceph-tools-54bdbfc7b7-c26ls                        1/1     Running     1 (127d ago)    136d    10.18.36.117    k8s-node1     <none>           <none>
rook-discover-hxtbj                                     1/1     Running     3 (127d ago)    146d    10.18.36.121    k8s-node1     <none>           <none>
rook-discover-m54kx                                     1/1     Running     6 (13d ago)     146d    10.18.159.178   k8s-master1   <none>           <none>
rook-discover-mxzfm                                     1/1     Running     2 (107d ago)    132d    10.18.224.36    k8s-master2   <none>           <none>
rook-discover-xv8xf                                     1/1     Running     3 (127d ago)    146d    10.18.135.198   k8s-master3   <none>           <none>





##############################################i try to apply csi-ceph-conf-override.yaml didn't work too########

[root@k8s-master2 ~]# kubectl -n rook-ceph get  configmap csi-ceph-conf-override -o yaml
apiVersion: v1
data:
  ceph.conf: |
    [global]
    rgw_thread_pool_size = 512
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"ceph.conf":"[global]\nrgw_thread_pool_size = 512\n"},"kind":"ConfigMap","metadata":{"annotations":{},"name":"csi-ceph-conf-override","namespace":"rook-ceph"}}
  creationTimestamp: "2023-09-13T09:14:29Z"
  name: csi-ceph-conf-override
  namespace: rook-ceph
  resourceVersion: "52299984"
  uid: 8f74541f-7114-40fc-b7ec-3c70762b899d
##############################################



[root@k8s-master2 ~]# kubectl -n rook-ceph get  configmap rook-ceph-csi-config -o yaml
apiVersion: v1
data:
  csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["10.15.202.8:6789","10.15.121.151:6789","10.15.217.97:6789"],"namespace":"rook-ceph"}]'
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-20T12:07:33Z"
  name: rook-ceph-csi-config
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: false
    controller: true
    kind: Deployment
    name: rook-ceph-operator
    uid: 16febb88-8f82-424c-b728-05f4b2f4945c
  resourceVersion: "52688105"
  uid: 275c7223-b826-4eb7-b304-ff7a7d99b6b1
[root@k8s-master2 ~]# kubectl -n rook-ceph get  configmap rook-ceph-mon-endpoints  -o yaml
apiVersion: v1
data:
  csi-cluster-config-json: '[{"clusterID":"rook-ceph","monitors":["10.15.121.151:6789","10.15.217.97:6789","10.15.202.8:6789"],"namespace":""}]'
  data: f=10.15.217.97:6789,e=10.15.202.8:6789,b=10.15.121.151:6789
  mapping: '{"node":{"b":{"Name":"k8s-master3","Hostname":"k8s-master3","Address":"10.10.5.162"},"e":{"Name":"k8s-node1","Hostname":"k8s-node1","Address":"10.10.5.163"},"f":{"Name":"k8s-master2","Hostname":"k8s-master2","Address":"10.10.5.161"}}}'
  maxMonId: "5"
  outOfQuorum: ""
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-20T12:07:36Z"
  finalizers:
  - ceph.rook.io/disaster-protection
  name: rook-ceph-mon-endpoints
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: ceph.rook.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: CephCluster
    name: rook-ceph
    uid: 41fe10f8-b09e-433b-b939-14e4e826489d
  resourceVersion: "52688061"
  uid: b23db2a8-e1d1-4919-a0dc-837cb862a2d8
[root@k8s-master2 ~]# 
[root@k8s-master2 ~]# 
[root@k8s-master2 ~]# 
[root@k8s-master2 ~]# 
[root@k8s-master2 ~]# kubectl -n rook-ceph get  configmap rook-ceph-operator-config  -o yaml
apiVersion: v1
data:
  CSI_CEPHFS_FSGROUPPOLICY: File
  CSI_ENABLE_CEPHFS_SNAPSHOTTER: "true"
  CSI_ENABLE_CSIADDONS: "false"
  CSI_ENABLE_ENCRYPTION: "false"
  CSI_ENABLE_LIVENESS: "false"
  CSI_ENABLE_NFS_SNAPSHOTTER: "true"
  CSI_ENABLE_RBD_SNAPSHOTTER: "true"
  CSI_ENABLE_TOPOLOGY: "false"
  CSI_FORCE_CEPHFS_KERNEL_CLIENT: "true"
  CSI_GRPC_TIMEOUT_SECONDS: "150"
  CSI_NFS_FSGROUPPOLICY: File
  CSI_PLUGIN_ENABLE_SELINUX_HOST_MOUNT: "false"
  CSI_PLUGIN_PRIORITY_CLASSNAME: system-node-critical
  CSI_PROVISIONER_PRIORITY_CLASSNAME: system-cluster-critical
  CSI_PROVISIONER_REPLICAS: "2"
  CSI_RBD_FSGROUPPOLICY: File
  ROOK_CEPH_ALLOW_LOOP_DEVICES: "false"
  ROOK_CEPH_COMMANDS_TIMEOUT_SECONDS: "15"
  ROOK_CSI_ALLOW_UNSUPPORTED_VERSION: "false"
  ROOK_CSI_ENABLE_CEPHFS: "true"
  ROOK_CSI_ENABLE_GRPC_METRICS: "false"
  ROOK_CSI_ENABLE_NFS: "false"
  ROOK_CSI_ENABLE_RBD: "true"
  ROOK_ENABLE_DISCOVERY_DAEMON: "true"
  ROOK_LOG_LEVEL: INFO
  ROOK_OBC_WATCH_OPERATOR_NAMESPACE: "true"
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-20T11:54:16Z"
  name: rook-ceph-operator-config
  namespace: rook-ceph
  resourceVersion: "44386"
  uid: 9ee70d6d-8137-4091-b7ae-dbd8562ea4c8
[root@k8s-master2 ~]# 







[root@k8s-master2 ~]# kubectl -n rook-ceph edit configmap rook-config-override


[root@k8s-master2 ~]# kubectl -n rook-ceph get  configmap rook-config-override -o yaml
apiVersion: v1
data:
  config: |
    [global]
    rgw thread pool size = 512
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-20T12:07:36Z"
  name: rook-config-override
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: ceph.rook.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: CephCluster
    name: rook-ceph
    uid: 41fe10f8-b09e-433b-b939-14e4e826489d
  resourceVersion: "52688011"
  uid: 94b84e67-a504-4f13-887a-61901ffa0c1a



configmap/rook-config-override edited
[root@k8s-master2 ~]# k delete po rook-ceph-rgw-my-store-a-6b8bfd75c6-s9fzl -n rook-ceph 
pod "rook-ceph-rgw-my-store-a-6b8bfd75c6-s9fzl" deleted
[root@k8s-master2 ~]# k delete po rook-ceph-rgw-pro-store-a-7cfbcb44f8-sw4sp -n rook-ceph 






[global]
rgw_thread_pool_size = 512	

[global]
rgw thread pool size = 512


[client]
rgw_thread_pool_size = 512

[client]
rgw thread pool size = 512


i have try all format


[root@k8s-master2 ~]#  k exec -it rook-ceph-tools-54bdbfc7b7-c26ls  bash -n rook-ceph2023/9/14

bash-4.4$ ceph config dump|grep rgw_thread_pool_size
bash-4.4$ ceph config dump|grep 'rgw thread pool size'
bash-4.4$ cat /etc/ceph/ceph.conf 
[global]
mon_host = 10.15.217.97:6789,10.15.202.8:6789,10.15.121.151:6789

[client.admin]
keyring = /etc/ceph/keyring




[root@k8s-master2 ~]# kubectl -n rook-ceph get  configmap rook-config-override -o yaml
apiVersion: v1
data:
  config: |
    [global]
    rgw thread pool size = 512
    [client]
    rgw_thread_pool_size = 512
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-20T12:07:36Z"
  name: rook-config-override
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: ceph.rook.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: CephCluster
    name: rook-ceph
    uid: 41fe10f8-b09e-433b-b939-14e4e826489d
  resourceVersion: "52699867"
  uid: 94b84e67-a504-4f13-887a-61901ffa0c1a

[root@k8s-master2 ~]# k delete po rook-ceph-rgw-my-store-a-6b8bfd75c6-jdlgl -n rook-ceph 
pod "rook-ceph-rgw-my-store-a-6b8bfd75c6-jdlgl" deleted
[root@k8s-master2 ~]# k delete po rook-ceph-rgw-pro-store-a-7cfbcb44f8-jvsk2 -n rook-ceph 
pod "rook-ceph-rgw-pro-store-a-7cfbcb44f8-jvsk2" deleted
[root@k8s-master2 ~]# 



[root@k8s-master2 ~]# kubectl -n rook-ceph get  configmap rook-config-override -o yaml
apiVersion: v1
data:
  ceph.conf: |
    [global]
    rgw_thread_pool_size = 512
  config: |
    [global]
    rgw thread pool size = 512
    [client.admin]
    rgw_thread_pool_size = 512
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-20T12:07:36Z"
  name: rook-config-override
  namespace: rook-ceph
  ownerReferences:
  - apiVersion: ceph.rook.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: CephCluster
    name: rook-ceph
    uid: 41fe10f8-b09e-433b-b939-14e4e826489d
  resourceVersion: "52701361"
  uid: 94b84e67-a504-4f13-887a-61901ffa0c1a

[root@k8s-master2 ~]# k delete po rook-ceph-rgw-my-store-a-6b8bfd75c6-sxlbp -n rook-ceph 
pod "rook-ceph-rgw-my-store-a-6b8bfd75c6-sxlbp" deleted
[root@k8s-master2 ~]# k delete po rook-ceph-rgw-pro-store-a-7cfbcb44f8-f4vjf -n rook-ceph 
pod "rook-ceph-rgw-pro-store-a-7cfbcb44f8-f4vjf" deleted
[root@k8s-master2 ~]# k delete po rook-ceph-tools-54bdbfc7b7-c26ls  -n rook-ceph
pod "rook-ceph-tools-54bdbfc7b7-c26ls" deleted


bash-4.4$ command terminated with exit code 137
[root@k8s-master2 ~]#  k exec -it rook-ceph-tools-54bdbfc7b7-whrb4  bash -n rook-ceph
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
bash-4.4$ cat /etc/ceph/ceph.conf 
[global]
mon_host = 10.15.217.97:6789,10.15.202.8:6789,10.15.121.151:6789

[client.admin]
keyring = /etc/ceph/keyring
bash-4.4$ 
bash-4.4$ ceph config dump|grep  rgw_thread_pool_size
bash-4.4$ ceph config dump|grep  'rgw\ thread\ pool\ size'
bash-4.4$ 







what default font-end is used in rook-ceph-rgw, it seems rgw_thread_pool_size parameter related to  CivetWeb or Beast


#############################
i am so helpless,what exactly  configmap i should edit and how to , then what pod should be deleted one by one to make it works

which format is right?

[root@k8s-master2 ~]# kubectl -n rook-ceph get  configmap rook-config-override -o yaml
apiVersion: v1
data:
  ceph.conf: |
    [global]
    rgw_thread_pool_size = 512
  config: |
    [global]
    rgw thread pool size = 512
    [client.admin]
    rgw_thread_pool_size = 512

**Reply by zhucan** [reply]
@wxcoin 
![image](https://github.com/rook/rook/assets/13997135/0a0545af-f653-4901-922e-07f6a93155af)
after restart the rgw pod, you can go into the  rgw pod, and run "ceph damon xxx.sock config ls | grep gw_thread_pool_size" to check the value.

**Reply by wxcoin** [reply]
thank you very much!
